src/Components/Browser.JS/src/Services/UriHelper.ts

```diff
@@ -1,69 +1,88 @@
+function listenForNavigationEvents(assemblyName: string, functionName: string) {
+  if (hasRegisteredNavigationEventListeners || assemblyName === undefined || functionName === undefined) {
     return;
   }
 
   notifyLocationChangedCallback = { assemblyName, functionName };
-  hasRegisteredEventListeners = true;
+
+  hasRegisteredNavigationEventListeners = true;
+  window.addEventListener('popstate', () => notifyLocationChanged(false));
+}
+
+function enableNavigationInterception() {
+  if (hasRegisteredNavigationInterception) {
+    return;
+  }
+
+  hasRegisteredNavigationInterception = true;
 
   document.addEventListener('click', event => {
+    if (event.button !== 0 || eventHasSpecialKey(event)) {
+      // Don't stop ctrl/meta-click (etc) from opening links in new tabs/windows
+      return;
+    }
+
     // Intercept clicks on all <a> elements where the href is within the <base href> URI space
     // We must explicitly check if it has an 'href' attribute, because if it doesn't, the result might be null or an empty string depending on the browser
     const anchorTarget = findClosestAncestor(event.target as Element | null, 'A') as HTMLAnchorElement;
     const hrefAttributeName = 'href';
-    if (anchorTarget && anchorTarget.hasAttribute(hrefAttributeName) && event.button === 0) {
-      const href = anchorTarget.getAttribute(hrefAttributeName)!;
-      const absoluteHref = toAbsoluteUri(href);
+    if (anchorTarget && anchorTarget.hasAttribute(hrefAttributeName)) {
       const targetAttributeValue = anchorTarget.getAttribute('target');
       const opensInSameFrame = !targetAttributeValue || targetAttributeValue === '_self';
+      if (!opensInSameFrame) {
+        return;
+      }
 
-      // Don't stop ctrl/meta-click (etc) from opening links in new tabs/windows
-      if (isWithinBaseUriSpace(absoluteHref) && !eventHasSpecialKey(event) && opensInSameFrame) {
+      const href = anchorTarget.getAttribute(hrefAttributeName)!;
+      const absoluteHref = toAbsoluteUri(href);
+
+      if (isWithinBaseUriSpace(absoluteHref)) {
         event.preventDefault();
-        performInternalNavigation(absoluteHref);
+        performInternalNavigation(absoluteHref, true);
       }
     }
   });
-
-  window.addEventListener('popstate', handleInternalNavigation);
 }
 
 export function navigateTo(uri: string, forceLoad: boolean) {
   const absoluteUri = toAbsoluteUri(uri);
 
   if (!forceLoad && isWithinBaseUriSpace(absoluteUri)) {
-    performInternalNavigation(absoluteUri);
+    performInternalNavigation(absoluteUri, false);
   } else {
     location.href = uri;
   }
 }
 
-function performInternalNavigation(absoluteInternalHref: string) {
+function performInternalNavigation(absoluteInternalHref: string, interceptedLink: boolean) {
   history.pushState(null, /* ignored title */ '', absoluteInternalHref);
-  handleInternalNavigation();
+  notifyLocationChanged(interceptedLink);
```

Comment:
Just to check, is there any chance that this will result in duplicate notifications?

I see that we're calling `pushState` above, which is good. But won't that also trigger the `popstate` listener we are also registering? If so it seems like `notifyLocationChanged` would get called twice.

If you could clarify what actually happens (and maybe add a comment, if this code is correct) that would be great.

Reply:
Oh wait, I misunderstood. Seems that [`pushState` does not trigger `popstate`](https://stackoverflow.com/a/26308536), so this would not cause duplicates.

Reply:
Yup. I misunderstood how `pushstate` works until I had a look at the documentation for it. It would have been very strange for it to raise the `popstate` event when you called `pushState` now that I think of it 😄